README.md
```
    _/      _/                                _/  _/
   _/_/  _/_/    _/_/_/  _/_/_/    _/_/_/        _/  _/
  _/  _/  _/  _/    _/  _/    _/  _/    _/  _/  _/_/
 _/      _/  _/    _/  _/    _/  _/    _/  _/  _/  _/
_/      _/    _/_/_/  _/_/_/    _/    _/  _/  _/    _/
                     _/
                    _/
```

[![Build Status Linux](https://api.travis-ci.org/CartoDB/mapnik.svg?branch=v3.0.15-carto)](http://travis-ci.org/CartoDB/mapnik)
[![Coverage Status](https://coveralls.io/repos/CartoDB/mapnik/badge.svg?branch=v3.0.15-carto&service=github)](https://coveralls.io/github/CartoDB/mapnik?branch=v3.0.15-carto)

Mapnik is an open source toolkit for developing mapping applications. At the core is a C++ shared library providing algorithms and patterns for spatial data access and visualization.

Mapnik is basically a collection of geographic objects like maps, layers, datasources, features, and geometries. The library doesn't rely on any OS specific "windowing systems" and it can be deployed to any server environment. It is intended to play fair in a multi-threaded environment and is aimed primarily, but not exclusively, at web-based development.

For further information see [http://mapnik.org](http://mapnik.org) and also our [wiki documentation](https://github.com/mapnik/mapnik/wiki).

# Installation

See [INSTALL.md](INSTALL.md) for installation instructions and the [Install](https://github.com/mapnik/mapnik/wiki/Mapnik-Installation) page on the wiki for guides.

# Code of Conduct

Please note that this project is released with a [Contributor Code of Conduct](https://github.com/mapnik/mapnik/blob/master/CODE_OF_CONDUCT.md). By participating in this project you agree to abide by its terms.

# License

Mapnik software is free and is released under the LGPL v2.1 ([GNU Lesser General Public License, version 2.1](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html)). Please see [COPYING](https://github.com/mapnik/mapnik/blob/master/COPYING) for more information.


debian/changelog
mapnik (3.0.18~rc1+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
  * Bump Standards-Version to 4.1.2, no changes.

 -- Bas Couwenberg <sebastic@debian.org>  Wed, 13 Dec 2017 18:40:54 +0100

mapnik (3.0.17+ds-1) unstable; urgency=medium

  * Move from experimental to unstable.

 -- Bas Couwenberg <sebastic@debian.org>  Thu, 30 Nov 2017 14:18:23 +0100

mapnik (3.0.17+ds-1~exp1) experimental; urgency=medium

  * New upstream release.
  * Drop scons-3.0.patch, fixed upstream.

 -- Bas Couwenberg <sebastic@debian.org>  Wed, 29 Nov 2017 20:54:24 +0100

mapnik (3.0.16+ds-1~exp1) experimental; urgency=medium

  * New upstream release.

 -- Bas Couwenberg <sebastic@debian.org>  Thu, 16 Nov 2017 18:44:31 +0100

mapnik (3.0.16~rc1+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
  * Drop icu-59.patch, fixed upstream.
  * Strip trailing whitespace from changelog.

 -- Bas Couwenberg <sebastic@debian.org>  Wed, 15 Nov 2017 21:12:32 +0100

mapnik (3.0.15+ds-3) unstable; urgency=medium

  * Add patch Matthias Klose by to fix build with ICU 59.1.
    (closes: #879835)

 -- Bas Couwenberg <sebastic@debian.org>  Thu, 26 Oct 2017 18:51:20 +0200

mapnik (3.0.15+ds-2) unstable; urgency=medium

  * Bump Standards-Version to 4.1.1, no changes.
  * Add patch to support scons 3.0.0.

 -- Bas Couwenberg <sebastic@debian.org>  Wed, 04 Oct 2017 19:54:53 +0200

mapnik (3.0.15+ds-1) unstable; urgency=medium

  * Bump Standards-Version to 4.0.0, no changes.
  * Add autopkgtest to test installability.
  * Move from experimental to unstable.

 -- Bas Couwenberg <sebastic@debian.org>  Tue, 29 Aug 2017 20:39:30 +0200

mapnik (3.0.15+ds-1~exp1) experimental; urgency=medium

  * New upstream release.

 -- Bas Couwenberg <sebastic@debian.org>  Fri, 16 Jun 2017 20:22:59 +0200

mapnik (3.0.15~rc1+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.

 -- Bas Couwenberg <sebastic@debian.org>  Thu, 15 Jun 2017 17:58:24 +0200

mapnik (3.0.14+ds-1~exp1) experimental; urgency=medium

  * New upstream release.

 -- Bas Couwenberg <sebastic@debian.org>  Mon, 05 Jun 2017 20:09:54 +0200

mapnik (3.0.14~rc1+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
  * Update copyright years for Artem Pavlenko.

 -- Bas Couwenberg <sebastic@debian.org>  Sat, 03 Jun 2017 10:25:05 +0200

mapnik (3.0.13+ds-1~exp2) experimental; urgency=medium

  * Merge changes from mapnik (3.0.12+ds-3).
    (closes: #859424)
  * Drop unused lintian overrides for hardening-no-pie.

 -- Bas Couwenberg <sebastic@debian.org>  Mon, 03 Apr 2017 15:36:12 +0200

mapnik (3.0.13+ds-1~exp1) experimental; urgency=medium

  * New upstream release.

 -- Bas Couwenberg <sebastic@debian.org>  Wed, 08 Feb 2017 18:21:29 +0100

mapnik (3.0.13~rc2+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
  * Require at least mapbox-variant 1.1.5.

 -- Bas Couwenberg <sebastic@debian.org>  Mon, 06 Feb 2017 18:57:41 +0100

mapnik (3.0.13~rc1+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
  * Re-instate lintian override for hardening-no-pie.
  * Drop dependencies-typo.patch, applied upstream.

 -- Bas Couwenberg <sebastic@debian.org>  Thu, 19 Jan 2017 00:11:00 +0100

mapnik (3.0.12+ds-3) unstable; urgency=medium

  * Update branch in gbp.conf & Vcs-Git URL.
  * Enable PIE hardening buildflags.
    (closes: #859424)

 -- Bas Couwenberg <sebastic@debian.org>  Mon, 03 Apr 2017 14:47:38 +0200

mapnik (3.0.12+ds-2) unstable; urgency=medium

  * Add patch to fix 'dependencies' typo.
  * Drop unused override for hardening-no-pie.
  * Change ttf-dejavu dependency to fonts-dejavu.
  * Reorder (build) dependencies.
  * Override dh_makeshlibs to use -V.
  * Add lintian override for no-symbols-control-file.

 -- Bas Couwenberg <sebastic@debian.org>  Sat, 05 Nov 2016 14:15:44 +0100

mapnik (3.0.12+ds-1) unstable; urgency=medium

  * New upstream release.
  * Update watchfile to not match releases/download.
  * Move from experimental to unstable.

 -- Bas Couwenberg <sebastic@debian.org>  Thu, 08 Sep 2016 15:27:37 +0200

mapnik (3.0.12~rc7+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
  * Update copyright file, changes:
    - Add license & copyright for travis-command-wrapper.py

 -- Bas Couwenberg <sebastic@debian.org>  Tue, 06 Sep 2016 11:16:42 +0200

mapnik (3.0.12~rc6+ds-1~exp2) experimental; urgency=medium

  * Add libcurl-ssl-dev as alternative to libcurl4-gnutls-dev dependencies.
    (closes: #836057)

 -- Bas Couwenberg <sebastic@debian.org>  Tue, 30 Aug 2016 12:59:34 +0200

mapnik (3.0.12~rc6+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.

 -- Bas Couwenberg <sebastic@debian.org>  Fri, 26 Aug 2016 18:56:15 +0200

mapnik (3.0.12~rc5+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.

 -- Bas Couwenberg <sebastic@debian.org>  Mon, 22 Aug 2016 22:21:22 +0200

mapnik (3.0.12~rc4+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
  * Drop Debian OpenStreetMap Team from Uploaders.

 -- Bas Couwenberg <sebastic@debian.org>  Thu, 18 Aug 2016 19:53:39 +0200

mapnik (3.0.12~rc3+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.

 -- Bas Couwenberg <sebastic@debian.org>  Fri, 12 Aug 2016 19:52:20 +0200

mapnik (3.0.12~rc2+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
  * Drop patches, applied upstream.

 -- Bas Couwenberg <sebastic@debian.org>  Wed, 10 Aug 2016 12:36:04 +0200

mapnik (3.0.12~rc1+ds-1~exp2) experimental; urgency=medium

  * Require at least mapbox-variant 1.1.1 for libmapnik-dev dependency too.

 -- Bas Couwenberg <sebastic@debian.org>  Sat, 06 Aug 2016 13:02:58 +0200

mapnik (3.0.12~rc1+ds-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
    - Fixes FTBFS with Boost 1.61.
  * Bump Standards-Version to 3.9.8, no changes.

src/build.py
#
# This file is part of Mapnik (c++ mapping toolkit)
#
# Copyright (C) 2015 Artem Pavlenko
#
# Mapnik is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
#


import os
import sys
import glob
from copy import copy
from subprocess import Popen, PIPE

Import('env')

lib_env = env.Clone()

def call(cmd, silent=True):
    stdin, stderr = Popen(cmd, shell=True, stdout=PIPE, stderr=PIPE).communicate()
    if not stderr:
        return stdin.strip()
    elif not silent:
        print (stderr)

def ldconfig(*args,**kwargs):
    call('ldconfig')

if env['LINKING'] == 'static':
    lib_env.Append(CXXFLAGS="-fPIC")

mapnik_lib_link_flag = ''

# note: .data gets the actual list to allow a true copy
# and avoids unintended pollution of other environments
libmapnik_cxxflags = copy(lib_env['CXXFLAGS'].data)
libmapnik_defines = copy(lib_env['CPPDEFINES'])

ABI_VERSION = env['ABI_VERSION']

enabled_imaging_libraries = []
filesystem = 'boost_filesystem%s' % env['BOOST_APPEND']
regex = 'boost_regex%s' % env['BOOST_APPEND']
system = 'boost_system%s' % env['BOOST_APPEND']

# clear out and re-set libs for this env
# note: order matters on linux: see lorder | tsort
lib_env['LIBS'] = [filesystem,
                   regex]

if env['COVERAGE']:
    lib_env.Append(LINKFLAGS='--coverage')
    lib_env.Append(CXXFLAGS='--coverage')

if env['HAS_CAIRO']:
    lib_env.Append(LIBS=env['CAIRO_ALL_LIBS'])

# maybe bz2
if len(env['EXTRA_FREETYPE_LIBS']):
    lib_env['LIBS'].extend(copy(env['EXTRA_FREETYPE_LIBS']))

if '-DHAVE_PNG' in env['CPPDEFINES']:
   lib_env['LIBS'].append('png')
   enabled_imaging_libraries.append('png_reader.cpp')

if '-DMAPNIK_USE_PROJ4' in env['CPPDEFINES']:
   lib_env['LIBS'].append('proj')

if '-DHAVE_TIFF' in env['CPPDEFINES']:
   lib_env['LIBS'].append('tiff')
   enabled_imaging_libraries.append('tiff_reader.cpp')

if '-DHAVE_WEBP' in env['CPPDEFINES']:
   lib_env['LIBS'].append('webp')
   enabled_imaging_libraries.append('webp_reader.cpp')

if env['XMLPARSER'] == 'libxml2' and env['HAS_LIBXML2']:
    lib_env['LIBS'].append('xml2')

if '-DBOOST_REGEX_HAS_ICU' in env['CPPDEFINES']:
    lib_env['LIBS'].append('icui18n')

lib_env['LIBS'].append(system)

lib_env['LIBS'].append('harfbuzz')

if '-DHAVE_JPEG' in env['CPPDEFINES']:
   lib_env['LIBS'].append('jpeg')
   enabled_imaging_libraries.append('jpeg_reader.cpp')

lib_env['LIBS'].append(env['ICU_LIB_NAME'])

lib_env['LIBS'].append('freetype')

if env['RUNTIME_LINK'] == 'static':
    if 'icuuc' in env['ICU_LIB_NAME']:
        lib_env['LIBS'].append('icudata')

if env['PLATFORM'] == 'Linux':
    lib_env['LINKFLAGS'].append('-pthread')

if env['RUNTIME_LINK'] != 'static':
    lib_env['LIBS'].insert(0, 'agg')

lib_env['LIBS'].append('z')

if env['PLATFORM'] == 'FreeBSD':
    lib_env['LIBS'].append('pthread')

if env['PLATFORM'] == 'Darwin':
    mapnik_libname = env.subst(env['MAPNIK_LIB_NAME'])
    if env['FULL_LIB_PATH']:
        lib_path = '%s/%s' % (env['MAPNIK_LIB_BASE'],mapnik_libname)
    else:
        lib_path = '@loader_path/'+mapnik_libname
    mapnik_lib_link_flag += ' -Wl,-install_name,%s' % lib_path
    _d = {'version':env['MAPNIK_VERSION_STRING'].replace('-pre','')}
    mapnik_lib_link_flag += ' -current_version %(version)s -compatibility_version %(version)s' % _d
else: # unix, non-macos
    mapnik_libname = env.subst(env['MAPNIK_LIB_NAME'])
    if env['ENABLE_SONAME']:
        mapnik_libname = env.subst(env['MAPNIK_LIB_NAME']) + (".%d.%d" % (int(ABI_VERSION[0]),int(ABI_VERSION[1])))
    if env['PLATFORM'] == 'SunOS':
        if env['CXX'].startswith('CC'):
            mapnik_lib_link_flag += ' -R. -h %s' % mapnik_libname
        else:
            mapnik_lib_link_flag += ' -Wl,-h,%s' %  mapnik_libname
    else: # Linux and others
        if env['PLATFORM'] != 'FreeBSD':
            lib_env['LIBS'].append('dl')
        mapnik_lib_link_flag += ' -Wl,-rpath-link,.'
        if env['ENABLE_SONAME']:
            mapnik_lib_link_flag += ' -Wl,-soname,%s' % mapnik_libname
        if env['FULL_LIB_PATH']:
            mapnik_lib_link_flag += ' -Wl,-rpath=%s' % env['MAPNIK_LIB_BASE']
        else:
            mapnik_lib_link_flag += ' -Wl,-z,origin -Wl,-rpath=\$$ORIGIN'

source = Split(
    """
    expression_grammar.cpp
    fs.cpp
    request.cpp
    well_known_srs.cpp
    params.cpp
    image_filter_types.cpp
    image_filter_grammar.cpp
    color.cpp
    conversions.cpp
    image_copy.cpp
    image_compositing.cpp
    image_scaling.cpp
    box2d.cpp
    datasource_cache.cpp
    datasource_cache_static.cpp
    debug.cpp
    geometry_reprojection.cpp
    expression_node.cpp
    expression_string.cpp
    expression.cpp
    transform_expression.cpp
    feature_kv_iterator.cpp
    feature_style_processor.cpp
    feature_type_style.cpp
    dasharray_parser.cpp
    font_engine_freetype.cpp
    font_set.cpp
    function_call.cpp
    gradient.cpp
    parse_path.cpp
    image_reader.cpp
    cairo_io.cpp
    image.cpp
    image_view.cpp
    image_view_any.cpp
    image_any.cpp
    image_options.cpp
    image_util.cpp
    image_util_jpeg.cpp
    image_util_png.cpp
    image_util_tiff.cpp
    image_util_webp.cpp
    layer.cpp
    map.cpp
    load_map.cpp
    palette.cpp
    marker_helpers.cpp
    transform_expression_grammar.cpp
    geometry_envelope.cpp
    plugin.cpp
    rule.cpp
    save_map.cpp
    wkb.cpp
    twkb.cpp
    projection.cpp
    proj_transform.cpp
    scale_denominator.cpp
    simplify.cpp
    parse_transform.cpp
    memory_datasource.cpp
    symbolizer.cpp
    symbolizer_keys.cpp
    symbolizer_enumerations.cpp
    unicode.cpp
    raster_colorizer.cpp
    mapped_memory_cache.cpp
    marker_cache.cpp
    svg/svg_parser.cpp
    svg/svg_path_parser.cpp
    svg/svg_points_parser.cpp
    svg/svg_transform_parser.cpp
    warp.cpp
    css_color_grammar.cpp
    vertex_cache.cpp
    vertex_adapters.cpp
    text/font_library.cpp
    text/text_layout.cpp
    text/text_line.cpp
    text/itemizer.cpp
    text/scrptrun.cpp
    text/face.cpp
    text/glyph_positions.cpp
    text/placement_finder.cpp
    text/properties_util.cpp
    text/renderer.cpp
    text/symbolizer_helpers.cpp
    text/text_properties.cpp
    text/font_feature_settings.cpp
    text/formatting/base.cpp
    text/formatting/list.cpp
    text/formatting/text.cpp
    text/formatting/format.cpp
    text/formatting/layout.cpp
    text/formatting/registry.cpp
    text/placements/registry.cpp
    text/placements/base.cpp
    text/placements/dummy.cpp
    text/placements/list.cpp
    text/placements/simple.cpp
    group/group_layout_manager.cpp
    group/group_rule.cpp
    group/group_symbolizer_helper.cpp
    xml_tree.cpp
    config_error.cpp
    color_factory.cpp
    renderer_common.cpp
    renderer_common/render_group_symbolizer.cpp
    renderer_common/render_markers_symbolizer.cpp
    renderer_common/render_pattern.cpp
    renderer_common/render_thunk_extractor.cpp
    math.cpp
    value.cpp
    metrics.cpp
    """
    )

if env['PLUGIN_LINKING'] == 'static':
    hit = False
    lib_env.AppendUnique(CPPPATH='../plugins/')
    for plugin in env['REQUESTED_PLUGINS']:
        details = env['PLUGINS'][plugin]
        if not details['lib'] or details['lib'] in env['LIBS']:
            plugin_env = SConscript('../plugins/input/%s/build.py' % plugin)
            if not plugin_env:
                print("Notice: no 'plugin_env' variable found for plugin: '%s'" % plugin)
            else:
                hit = True
                DEF = '-DMAPNIK_STATIC_PLUGIN_%s' % plugin.upper()
                lib_env.Append(CPPDEFINES = DEF)
                if DEF not in libmapnik_defines:
                    libmapnik_defines.append(DEF)
                if 'SOURCES' in plugin_env and plugin_env['SOURCES']:
                    source += ['../plugins/input/%s/%s' % (plugin, src) for src in plugin_env['SOURCES']]
                if 'CPPDEFINES' in plugin_env and plugin_env['CPPDEFINES']:
                    lib_env.AppendUnique(CPPDEFINES=plugin_env['CPPDEFINES'])
                if 'CXXFLAGS' in plugin_env and plugin_env['CXXFLAGS']:
                    lib_env.AppendUnique(CXXFLAGS=plugin_env['CXXFLAGS'])
                if 'LINKFLAGS' in plugin_env and plugin_env['LINKFLAGS']:
                    lib_env.AppendUnique(LINKFLAGS=plugin_env['LINKFLAGS'])
                if 'CPPPATH' in plugin_env and plugin_env['CPPPATH']:
                    lib_env.AppendUnique(CPPPATH=copy(plugin_env['CPPPATH']))
                if 'LIBS' in plugin_env and plugin_env['LIBS']:
                    lib_env.AppendUnique(LIBS=plugin_env['LIBS'])
        else:
            print("Notice: dependencies not met for plugin '%s', not building..." % plugin)
    if hit:
        lib_env.Append(CPPDEFINES = '-DMAPNIK_STATIC_PLUGINS')
        libmapnik_defines.append('-DMAPNIK_STATIC_PLUGINS')

# add these to the compile flags no matter what
# to make it safe to try to compile them from Makefile wrapper
source += Split("""
cairo/process_markers_symbolizer.cpp
cairo/process_group_symbolizer.cpp
""")

if env['HAS_CAIRO']:
    lib_env.AppendUnique(LIBPATH=env['CAIRO_LIBPATHS'])
    lib_env.Append(CPPDEFINES = '-DHAVE_CAIRO')
    libmapnik_defines.append('-DHAVE_CAIRO')
    lib_env.AppendUnique(CPPPATH=copy(env['CAIRO_CPPPATHS']))
    source += Split("""
    cairo/cairo_context.cpp
    cairo/cairo_renderer.cpp
    cairo/cairo_render_vector.cpp
    cairo/process_text_symbolizer.cpp
    cairo/process_line_symbolizer.cpp
    cairo/process_line_pattern_symbolizer.cpp
    cairo/process_polygon_symbolizer.cpp
    cairo/process_polygon_pattern_symbolizer.cpp
    cairo/process_debug_symbolizer.cpp
    cairo/process_point_symbolizer.cpp
    cairo/process_raster_symbolizer.cpp
    cairo/process_building_symbolizer.cpp
    """)

for cpp in enabled_imaging_libraries:
    source.append(cpp)

# agg backend
source += Split(
    """
    agg/agg_renderer.cpp
    agg/process_dot_symbolizer.cpp
    agg/process_building_symbolizer.cpp
    agg/process_line_symbolizer.cpp
    agg/process_line_pattern_symbolizer.cpp
    agg/process_text_symbolizer.cpp
    agg/process_point_symbolizer.cpp
    agg/process_polygon_symbolizer.cpp
    agg/process_polygon_pattern_symbolizer.cpp
    agg/process_raster_symbolizer.cpp
    agg/process_shield_symbolizer.cpp
    agg/process_markers_symbolizer.cpp
    agg/process_group_symbolizer.cpp
    agg/process_debug_symbolizer.cpp
    """
    )

if env['RUNTIME_LINK'] == "static":
    source += glob.glob('../deps/agg/src/' + '*.cpp')

# add these to the compile flags no matter what
# to make it safe to try to compile them from Makefile wrapper
source += Split("""
grid/process_markers_symbolizer.cpp
grid/process_group_symbolizer.cpp
""")

# grid backend
if env['GRID_RENDERER']:
    source += Split(
        """
        grid/grid.cpp
        grid/grid_renderer.cpp
        grid/process_building_symbolizer.cpp
        grid/process_line_pattern_symbolizer.cpp
        grid/process_line_symbolizer.cpp
        grid/process_point_symbolizer.cpp
        grid/process_polygon_pattern_symbolizer.cpp
        grid/process_polygon_symbolizer.cpp
        grid/process_raster_symbolizer.cpp
        grid/process_shield_symbolizer.cpp
        grid/process_text_symbolizer.cpp
        """)
    lib_env.Append(CPPDEFINES = '-DGRID_RENDERER')
    libmapnik_defines.append('-DGRID_RENDERER')

# https://github.com/mapnik/mapnik/issues/1438
if env['SVG_RENDERER']: # svg backend
    source += Split(
    """
    svg/output/svg_output_grammars.cpp
    svg/output/svg_renderer.cpp
    svg/output/svg_generator.cpp
    svg/output/svg_output_attributes.cpp
    svg/output/process_symbolizers.cpp
    svg/output/process_line_symbolizer.cpp
    svg/output/process_polygon_symbolizer.cpp
    """)
    lib_env.Append(CPPDEFINES = '-DSVG_RENDERER')
    libmapnik_defines.append('-DSVG_RENDERER')

if env['XMLPARSER'] == 'libxml2' and env['HAS_LIBXML2']:
    source += Split(
        """
        libxml2_loader.cpp
        """)
    lib_env.Append(CPPDEFINES = '-DHAVE_LIBXML2')
    libmapnik_defines.append('-DHAVE_LIBXML2')
else:
    source += Split(
        """
        rapidxml_loader.cpp
        """
    )

# clone the env one more time to isolate mapnik_lib_link_flag
lib_env_final = lib_env.Clone()
lib_env_final.Prepend(LINKFLAGS=mapnik_lib_link_flag)

# cache library values for other builds to use
env['LIBMAPNIK_LIBS'] = copy(lib_env['LIBS'])
env['LIBMAPNIK_LINKFLAGS'] = copy(lib_env['LINKFLAGS'])
env['LIBMAPNIK_CXXFLAGS'] = libmapnik_cxxflags
env['LIBMAPNIK_DEFINES'] = libmapnik_defines

mapnik = None

if env['PLATFORM'] == 'Darwin' or not env['ENABLE_SONAME']:
    target_path = env['MAPNIK_LIB_BASE_DEST']
    if 'uninstall' not in COMMAND_LINE_TARGETS:
        if env['LINKING'] == 'static':
            mapnik = lib_env_final.StaticLibrary(env['MAPNIK_NAME'], source)
        else:
            mapnik = lib_env_final.SharedLibrary(env['MAPNIK_NAME'], source)
        result = env.Install(target_path, mapnik)
        env.Alias(target='install', source=result)

    env['create_uninstall_target'](env, os.path.join(target_path,env.subst(env['MAPNIK_LIB_NAME'])))
else:
    # Symlink command, only works if both files are in same directory
    def symlink(env, target, source):
        trgt = str(target[0])
        src = str(source[0])

        if os.path.islink(trgt) or os.path.exists(trgt):
            os.remove(trgt)
        os.symlink(os.path.basename(src), trgt)

    major, minor, micro = ABI_VERSION

    soFile = "%s.%d.%d.%d" % (os.path.basename(env.subst(env['MAPNIK_LIB_NAME'])), int(major), int(minor), int(micro))
    target = os.path.join(env['MAPNIK_LIB_BASE_DEST'], soFile)

    if 'uninstall' not in COMMAND_LINE_TARGETS:
        if env['LINKING'] == 'static':
            mapnik = lib_env_final.StaticLibrary(env['MAPNIK_NAME'], source)
        else:
            mapnik = lib_env_final.SharedLibrary(env['MAPNIK_NAME'], source)
        result = env.InstallAs(target=target, source=mapnik)
        env.Alias(target='install', source=result)
        if result:
              env.AddPostAction(result, ldconfig)

    # Install symlinks
    target1 = os.path.join(env['MAPNIK_LIB_BASE_DEST'], "%s.%d.%d" % \
        (os.path.basename(env.subst(env['MAPNIK_LIB_NAME'])),int(major), int(minor)))
    target2 = os.path.join(env['MAPNIK_LIB_BASE_DEST'], os.path.basename(env.subst(env['MAPNIK_LIB_NAME'])))
    if 'uninstall' not in COMMAND_LINE_TARGETS:
        link1 = env.Command(target1, target, symlink)
        env.Alias(target='install', source=link1)
        link2 = env.Command(target2, target1, symlink)
        env.Alias(target='install', source=link2)
    # delete in reverse order..
    env['create_uninstall_target'](env, target2)
    env['create_uninstall_target'](env, target1)
    env['create_uninstall_target'](env, target)

    # to enable local testing
    lib_major_minor = "%s.%d.%d" % (os.path.basename(env.subst(env['MAPNIK_LIB_NAME'])), int(major), int(minor))
    local_lib = os.path.basename(env.subst(env['MAPNIK_LIB_NAME']))
    if os.path.islink(lib_major_minor) or os.path.exists(lib_major_minor):
        os.remove(lib_major_minor)
    os.symlink(local_lib,lib_major_minor)
    Clean(mapnik,lib_major_minor);

if not env['RUNTIME_LINK'] == 'static':
    Depends(mapnik, env.subst('../deps/agg/libagg.a'))
